src/main/kotlin/org/rust/cargo/project/model/impl/CargoProjectImpl.kt

```diff
@@ -0,0 +1,378 @@
+    }
+
+    /**
+     * All modifications to project model except for low-level `loadState` should
+     * go through this method: it makes sure that when we update various IDEA listeners,
+     * [allProjects] contains fresh projects.
+     */
+    private fun modifyProjects(
+        f: (List<CargoProjectImpl>) -> Promise<List<CargoProjectImpl>>
+    ): Promise<List<CargoProjectImpl>> =
+        projects.updateAsync(f)
+            .then { projects ->
+                directoryIndex.resetIndex()
+                ApplicationManager.getApplication().invokeAndWait {
+                    runWriteAction {
+                        ProjectRootManagerEx.getInstanceEx(project)
+                            .makeRootsChange(EmptyRunnable.getInstance(), false, true)
+                    }
+                }
+                project.messageBus.syncPublisher(CargoProjectsService.CARGO_PROJECTS_TOPIC)
+                    .cargoProjectsUpdated(projects)
+
+                projects
+            }
+
+    @TestOnly
+    override fun createTestProject(rootDir: VirtualFile, ws: CargoWorkspace) {
+        val manifest = rootDir.pathAsPath.resolve("Cargo.toml")
+        val testProject = CargoProjectImpl(manifest, this, ws, null, UpdateStatus.UpToDate)
+        testProject.setRootDir(rootDir)
+        modifyProjects { _ ->
+            Promise.resolve(listOf(testProject))
+        }
+            .blockingGet(1, TimeUnit.MINUTES)
+    }
+
+    override fun getState(): Element {
+        val state = Element("state")
+        for (cargoProject in allProjects) {
+            val cargoProjectElement = Element("cargoProject")
+            cargoProjectElement.setAttribute("FILE", cargoProject.manifest.systemIndependentPath)
+            state.addContent(cargoProjectElement)
+        }
+        return state
+    }
+
+    override fun loadState(state: Element) {
+        val loaded = state.getChildren("cargoProject")
+            .mapNotNull { it.getAttributeValue("FILE") }
+            .map { CargoProjectImpl(Paths.get(it), this) }
+        // Refresh projects via `invokeLater` to avoid model modifications
+        // while the project is being opened. Use `updateSync` directly
+        // instead of `modifyProjects` for this reason
+        projects.updateSync { _ -> loaded }
+            .done {
+                ApplicationManager.getApplication().invokeLater { refreshAllProjects() }
+            }
+    }
+
+    override fun noStateLoaded() {
+        // Do nothing: in theory, we might try to do [discoverAndRefresh]
+        // here, but the `RustToolchain` is most likely not ready.
+        //
+        // So the actual "Let's guess a project model if it is not imported
+        // explicitly" happens in [org.rust.ide.notifications.MissingToolchainNotificationProvider]
+    }
+
+    override fun toString(): String {
+        return "CargoProjectsService(projects = $allProjects)"
+    }
+}
+
+data class CargoProjectImpl(
```

Comment:
Extra file?

Reply:
No, I intentionally have separated interface and `Impl` classes. Although in most cases it is stupid to have an interface with a single implementation only, I think it makes sense in this particular use-case, because the interface is rather important and should be used throughout the project, and the impl is quite involved.